### DISCUSSION TITLE
OpenDTU direkt mit Tasmota steuern über API

### DISCUSSION THREAD

**Discussion by johannesosterberg** [tbnobody/OpenDTU#234] [Q&A] [answered]
Hi, ich probiere gerade eine Nulleinspeisung über einen Tasmota (Volkszähler) der am Stromzähler hängt, zu realisieren.

Tasmota kann [**WebQuery**](https://tasmota.github.io/docs/Commands/#wi-fi), mit denen man über die [**Rules**](https://tasmota.github.io/docs/Rules/) so etwas steuern könnt.

Ich stehe gerade nur auf dem Schlauch, wie ich über die Post Methode das Limiter des Inverters steuern kann. 

Es müsste ja mehr oder weniger mit http://ip/api/limit/config gehen, wird das dann mittels JSON übertragen oder wie funktioniert das genau?

Habe ein wenig mit **curl** herumexperimentiert, aber komme nicht weiter.

curl -d '{"11xxxxxxxx":{"limit_relative“:50}}' -H "Content-Type: application/json" -X POST  http://ip/api/limit/config

Bringt mich auf jeden Fall nicht weiter. Hat jemand da eine Lösung für.

Vielen lieben Dank. Grüße.

**Comment by PeterFausH** [comment, marked as answer]
Der Volkszähler hat ein miniweb das du in der .conf auf Port 8080 aktivieren kannst. Dort kannst du auf JSON-Daten zugreifen. 
Mit einem Python-Script hole ich mir da den Hausverbrauch und setze dann den Limit absolut auf den aktuellen (meist niedrigeren) Verbrauch. Du musst aber ein Delta aus altem Limit und aktuellem Verbrauchswert errechnen da dein Balkonkraftwerk ja den Bezug schon reduziert hat.
`# Zugang zum Volkszähler MiniWeb`
`apiURL = "http://192.168.200.207:8080"  # miniweb des Volkszähler im Chalet`

Um den aktuellen Bezug festzustellen nehme ich:
`# Bezug vom Chalet feststellen`
`r = requests.get(apiURL)  # Volkszähler lesen`
`json_string = r.text`
`parsed_json = json.loads(json_string)`
`# hier anpassen an json Ausgabe`
`grid_sum = int(parsed_json['data'][1]['tuples'][0][1])`
`print("aktueller Bezug Chalet:   ",grid_sum)`

Hilft dir das?

**Reply by johannesosterberg** [reply]
Vielen Dank, ich habe es momentan mit Python Skript mit MQTT gelöst. 

Prinzipiell wollte ich das mit nur zwei Geräten lösen. 
Einmal den EPS32 mit OpenDTU und den Volkszähler/Tasmota ESP, die sollen sich untereinander abstimmen, wie viel Strom eingespeist werden soll.

Jetzt habe ich halt noch einen MQTT Broker, der auf meinem Webserver läuft dazwischen. Vielleicht sollte ich am OpenDTU ein wenig rumwerkeln, dass der den Volkszähler direkt auslesen kann und sich dann entsprechend einstellt.

Ich habe auf Github ein Proof of Concept eines Python-Skriptes gestellt, vielleicht hilft das dem einen oder anderen.
Ich bin aber kein Programmierer und es können Fehler enthalten sein! :)

[hm_nulleinspeisung](https://github.com/johannesosterberg/hm_nulleinspeisung)

**Reply by x90cr** [reply]
Ich habe heute auch den ganzen Tag dran gesessen und da ich sonst nirgends eine Lösung gefunden haben möchte ich hier für andere Suchende wenigstens die Lösung da lassen:

Vor dem http request Node macht ihr ein function Node mit folgendem Code:

msg.headers = {'Content-Type': 'application/x-www-form-urlencoded'}
    
// payload-Objekt erstellen
var payloadObj = {
    "serial": "1141749xxxxx",
    "limit_type": 1,
    "limit_value": 50
};

// Payload in das richtige Format umwandeln
msg.payload = "data=" + JSON.stringify(payloadObj);


return msg;
